rmw_fastrtps_dynamic_cpp/src/listener_thread.cpp

```diff
@@ -0,0 +1,181 @@
+    return RMW_RET_ERROR;
+  } catch (...) {
+    RMW_SET_ERROR_MSG("Failed to join std::thread");
+    return RMW_RET_ERROR;
+  }
+  rmw_ret = rmw_destroy_guard_condition(common_context->listener_thread_gc);
+  if (RMW_RET_OK != rmw_ret) {
+    return rmw_ret;
+  }
+  return RMW_RET_OK;
+}
+
+static
+void
+terminate(const char * error_message)
+{
+  RCUTILS_LOG_ERROR_NAMED(log_tag, "%s, terminating ...", error_message);
+  std::terminate();
+}
+
+void
+node_listener(rmw_context_t * context)
+{
+  assert(nullptr != context);
+  assert(nullptr != context->impl);
+  assert(nullptr != context->impl->common);
+  auto common_context = static_cast<rmw_dds_common::Context *>(context->impl->common);
+  while (common_context->thread_is_running.load()) {
+    assert(nullptr != common_context->sub);
+    assert(nullptr != common_context->sub->data);
+    void * subscriptions_buffer[] = {common_context->sub->data};
+    void * guard_conditions_buffer[] = {common_context->listener_thread_gc->data};
+    rmw_subscriptions_t subscriptions;
+    rmw_guard_conditions_t guard_conditions;
+    subscriptions.subscriber_count = 1;
+    subscriptions.subscribers = subscriptions_buffer;
+    guard_conditions.guard_condition_count = 1;
+    guard_conditions.guard_conditions = guard_conditions_buffer;
+    // number of conditions of a subscription is 2
+    rmw_wait_set_t * wait_set = rmw_create_wait_set(context, 2);
+    if (nullptr == wait_set) {
+      terminate("failed to create wait set");
+    }
+    if (RMW_RET_OK != rmw_wait(
+        &subscriptions,
+        &guard_conditions,
+        nullptr,
+        nullptr,
+        nullptr,
+        wait_set,
+        nullptr))
+    {
+      terminate("rmw_wait failed");
+    }
+    if (subscriptions_buffer[0]) {
+      rmw_dds_common::msg::ParticipantEntitiesInfo msg;
+      bool taken;
+      if (RMW_RET_OK != rmw_take(
+          common_context->sub,
+          static_cast<void *>(&msg),
+          &taken,
+          nullptr))
+      {
+        terminate("rmw_take failed");
+      }
+      if (taken) {
+        // TODO(ivanpauno): Should the program be terminated if taken is false?
+        if (std::memcmp(
+            reinterpret_cast<char *>(common_context->gid.data),
+            reinterpret_cast<char *>(&msg.gid.data),
+            RMW_GID_STORAGE_SIZE) == 0)
+        {
+          // ignore local messages
+          continue;
+        }
+        common_context->graph_cache.update_participant_entities(msg);
+        if (rcutils_logging_logger_is_enabled_for("rmw_dds_common",
+          RCUTILS_LOG_SEVERITY_DEBUG))
+        {
+          std::ostringstream ss;
+          ss << common_context->graph_cache;
+          RCUTILS_LOG_DEBUG_NAMED("rmw_fastrtps_dynamic_cpp", "%s", ss.str().c_str());
+        }
+      }
+    }
+    if (RMW_RET_OK != rmw_destroy_wait_set(wait_set)) {
+      terminate("failed to destroy wait set");
+    }
+  }
+}
```

Comment:
The problem is that I'm using `rmw` api directly in `listener_thread.cpp`, so I can't move it to shared.

Reply:
But this part of the API is implemented in shared anyways right? You could use the shared API instead of the rmw API maybe. That would let you move it. I agree this is a lot of duplication.

Reply:
Moved to `rmw_fastrtps_shared_cpp`. I also did the same with some other duplicated code.

I couldn't do that in other places, e.g.: `rmw_init` calls `create_publisher`, which is different for each kind of typesupport.

----

Bottom line: AFAIS, only a small part of the code is different in each rmw implementation. Particularly, how the typesupport is created/registered when creating a publisher/subscription/client/service.
If it's just that, wouldn't make more sense to have just one implementation and use macros to compile both libraries from the same source?
I think that that solves the dependency inversion problem (because there's not `rmw_fastrtps_shared_cpp` library at all), it avoids boilerplate code (wrapping the `rmw_fastrtps_shared_cpp` function), and it makes the code more easy to maintain (about half the lines of code).
If we want to have one package for each of the libraries, we could still have the source code in a third package, but which just "exports" the location of the source without building the library (or another similar trick).

Sorry for the unrelated comment, I just wanted to share my thoughts about this problem.

Reply:
Those both sound like reasonable things to do. Also, we could consider getting rid of dynamic, if we have another set of things that use the introspection type support.

But all of that is out of scope for this one, I guess. Thanks for trying to de-duplicate more.